Review

konopat opened the pull request.

konopat committed:
chore: Add .DS_Store to .gitignore

konopat committed:
feat: Add ingredients to Redux store

konopat committed:
feat: Add base routing and loading

konopat committed:
feat: adding ingredients to the burger constructor

konopat committed:
feat: counting ingredients, remove/move an ingredient in the burger constructor

konopat committed:
feat: protected route

konopat committed:
feat: button navigation

konopat committed:
feat: register/login with api

konopat committed:
feat(refactoring): working with Redux

konopat committed:
feat: user state

konopat committed:
feat: use alias @utils-cookie

konopat committed:
feat: correct user profile

konopat committed:
feat: just tidied up

aromanovnik commented on src/components/ui/app-header/app-header.tsx:
```diff
@@ -7,28 +7,49 @@ import {
+        <NavLink
+          to='/profile'
+          className={clsx(styles.link, {
+            [styles.link_active]: location.pathname === '/profile'
```
**Надо исправить**: Ссылка должна оставаться активной, если пользователь находится на вложенных  маршрутах.  Рекомендуется использовать "dynamic className" - когда `className` передается как функция, которая принимает параметры и возвращает строку с классами.

<img width="1316" alt="Screenshot 2024-11-22 at 15 40 18" src="https://github.com/user-attachments/assets/66ca706d-071d-428e-a899-1e3e1f71ae5d">

aromanovnik commented on src/components/protected-route/index.tsx:
```diff
@@ -0,0 +1,27 @@
+import { Preloader } from '@ui';
+import { IProtectedRouteProps } from '@utils-types';
+
+export const ProtectedRoute: FC<IProtectedRouteProps> = ({
```
**Надо исправить**: Роуты `/login`, `/register`, `/forgot-password` и `/reset-password` должны быть недоступны для авторизованных пользователей. Авторизованный пользователь при попытке перейти на эти страницы должен быть перенаправлен на главную страницу или страницу профиля.

<img width="1131" alt="Screenshot 2024-11-22 at 15 50 43" src="https://github.com/user-attachments/assets/747861ac-e68d-4500-a01e-657da8d77802">

aromanovnik commented on src/components/protected-route/index.tsx:
```diff
@@ -0,0 +1,27 @@
+    return <Preloader />;
+  }
+
+  // Если маршрут защищён и пользователь не авторизован
```
**Надо исправить**: После успешной авторизации должно происходить перенаправление на тот маршрут, который пользователь пытался посетить ранее.

aromanovnik commented on src/components/protected-route/index.tsx:
```diff
@@ -0,0 +1,27 @@
+  isProtected = true
+}) => {
+  const location = useLocation();
+  const { user, isChecked, isLoading } = useAppSelector(
```
**Можно лучше**: Правильнее делать отдельные селекторы, чтобы изолировать логику доступа к данным состояния и избежать дублирования кода при доступе к данным в разных частях приложения. Это также позволяет легче изменять структуру состояния в будущем без необходимости вносить изменения во множество компонентов.

konopat replied:
Спасибо за подсказку с селекторами! Исправил, вроде, везде. Код стал действительно понятнее.

aromanovnik reviewed (changes requested):
Добрый день, Роман!

Работа выполнена очень хорошо: конструктор бургеров работает без сбоев, оформить заказ может только авторизованный пользователь, запросы к серверу корректно выполняются через хук useEffect, управление модальными окнами — как с данными ингредиента, так и с информацией о заказах из ленты и истории заказов — реализовано с использованием роутинга. Видно хорошее понимание Redux, код выглядит чисто и аккуратно.

Тем не менее, есть несколько замечаний, которые потребуют доработки. Более подробно о них написано в комментариях к коду.

Что нужно исправить:
* Роуты `/login`, `/register`, `/forgot-password` и `/reset-password` должны быть недоступны для авторизованных пользователей. Авторизованный пользователь при попытке перейти на эти страницы должен быть перенаправлен на главную страницу или страницу профиля.
* После успешной авторизации должно происходить перенаправление на тот маршрут, который пользователь пытался посетить ранее.
* Ссылка должна оставаться активной, если пользователь находится на вложенных маршрутах. 

Что можно улучшить:
* Правильнее делать отдельные селекторы, чтобы изолировать логику доступа к данным состояния и избежать дублирования кода при доступе к данным в разных частях приложения.

Напоминаю, что работа может быть принята только после исправления всех критических замечаний **Надо исправить**.

Удачного рефакторинга кода.

konopat committed:
fix: profile link stay active when user is inside

konopat committed:
fix: redirecting authorized users from routes: login, /registration, /forgotten password and /password reset

konopat committed:
fix: redirect to request before authorization

konopat committed:
feat: selectors to work with states

aromanovnik reviewed (approved):
Поздравляю! Ваша работа принята. Все недочеты исправлены. Желаю успехов в дальнейшей учебе!

konopat committed:
test: добавлены e2e тесты с Cypress

- Добавлены тесты для модального окна ингредиента (открытие/закрытие)
- Добавлены тесты для добавления ингредиентов в конструктор
- Добавлены тесты для создания заказа
- Настроено тестовое окружение Cypress
- Добавлены моковые данные для тестов

konopat committed:
build: установка и настройка Cypress

- Установлен Cypress как dev-зависимость
- Добавлена базовая конфигурация cypress.config.js
- Настроены вспомогательные файлы в cypress/support
- Добавлены скрипты запуска в package.json

konopat committed:
test: добавлены unit тесты для Redux store

- Добавлены тесты для проверки инициализации rootReducer
- Добавлены тесты для constructor-slice (добавление, удаление, перемещение)
- Добавлены тесты для ingredients-slice (загрузка данных)
- Добавлены тесты для user-slice (авторизация)
- Добавлены тесты для orders-slice (создание заказа)
- Настроена конфигурация Jest
- Добавлены необходимые моки
